### ISSUE TITLE
[display] read-only display requests invalidate module

### ISSUE THREAD

**Issue by nadako** [HaxeFoundation/haxe#8596]
(even though it's a big topic code-wise, i'm just creating the issue so we don't forget about it as it's quite important for any large code-base)

If I understand correctly, all the display requests, including the pure "read-only" ones like hover, position, type position, etc currently invalidate the module and all its dependencies.

This is very unfortunate, because mere code navigation on a large code-base with a lot of inter-dependent modules makes the completion server unusably slow very quickly. Especially if you navigate to and do some hovering in some basic building block like `flash.events.Event` for Flash/OpenFL projects.

PS I expect this to be a big issue preventing us from using VS Code when we move to the Haxe development in InnoGames so this is something that might be requested to look into by "us" then.

**Comment by Simn** [comment]
I agree that we need this, but it's tricky to implement. We want to force typing of modules without removing them from the cache.

The simplest approach might be to remove them from the cache, trigger the typing and then add them back in the end. This should be safe because we don't update the cache on display requests anyway.

**Comment by nadako** [comment]
"The simplest approach" doesn't sound that tricky :)

**Comment by Simn** [comment]
Well I came up with that while writing the first sentence. :P

It might be straightforward, or it might be a huge mess due to something we're not considering.

**Comment by Simn** [comment]
See if that does anything for you. I tested it for 30 seconds and it didn't immediately break displaying in my Hello World program.

**Comment by nadako** [comment]
it doesn't break stuff for me and i _think_ it's better, but i still see some module skipping after jumping between some modules. can't yet easily reproduce though :(

**Comment by nadako** [comment]
> it doesn't break stuff

ok it does, I'm getting some new "module is redefined" errors. still not quite sure where it comes from

**Comment by nadako** [comment]
This happens with two inter-dependent modules that are both included in the compilation, e.g:

```haxe
import B;

class A {}
```

```haxe
import A;

class B {}
```

(just add both `A` and `B` in the hxml)

Hovering either A or B will result in `Type name A is redefined from module A` (or `B`).

**Comment by Simn** [comment]
Please try again now. I've added your mutually dependent module situation as a test so we can make sure that it doesn't regress again.

I'm still not 100% comfortable with this approach but it's quite hard to judge due to the complexity of the compilation server.

**Comment by nadako** [comment]
Thanks! It's amazing that we have tests for this stuff nowadays!

Will do some testing on our codebase soon

**Comment by nadako** [comment]
Everything seems to be fine so far! Don't see any tainted modules in the `-v` output while navigating \o/

**Comment by nadako** [comment]
Found one issue with diagnostics, with interdependent modules again:

```haxe
class Other {
	public static function f(m:Main) {}
}
```

```haxe
class Main {
	function new() {}

	static function main() {
		Other.f(new Main());
	}
}
```

for the `new Main()` diagnostics report: `Main should be Main`. Compilation and everything else seem to work fine still.

**Comment by Simn** [comment]
Right, shit, I think this means that the entire approach isn't sound. The `Other` module has a reference to the original `Main` in its AST, and by re-typing `Main` during the diagnostics run we end up with two different `Main` modules.

This probably only shows up in diagnostics because that's the only mode where we actually care about the unification failing. The problem itself would still be present in other modes though.

That's quite unfortunate because I can't see how we would make this work right now. Will think about it some more, but we might have to revert.

**Comment by nadako** [comment]
If it cannot mess up the cache I would vote for not reverting it even if it's not easy to fix or not the right approach in general, because well diagnostics can be disabled, but working completion and navigation is very important. Maybe we could hide it with a define or cache server configuration flag...

**Comment by lublak** [comment]
I have a big project and have hundreds of these messages. 
Troubleshooting has become uglier.
And I think it should not be the thought of having such error messages.

**Comment by Simn** [comment]
Have to revert this for the 4.0 branch.

**Comment by RealyUniqueName** [comment]
Reverted in 4.0. Moving to 4.1

**Comment by Simn** [comment]
This has been addressed in #8939, with the remaining problems being tracked in #8949 and across various individual issues @nadako comes across.
